### ISSUE TITLE
SNOW-872164: TypeError: Cannot read properties of null (reading 'getPeerCertificate') resulting in S3/Blob read error

### ISSUE THREAD

**Issue by kumaresansv** [snowflakedb/snowflake-connector-nodejs#588]
Please answer these questions before submitting your issue. 
In order to accurately debug the issue this information is required. Thanks!

1. What version of NodeJS driver are you using?
1.6.22 & 1.6.23 (Tried with both)
   
2. What operating system and processor architecture are you using?
Linux 
   
3. What version of NodeJS are you using?
node:20-alpine

4. What are the component versions in the environment (`npm list`)?
up-report-generation@1.0.0 /Users/sankaranak/code/qp-report-generation
├── @google-cloud/secret-manager@4.2.2
├── @google-cloud/storage@6.12.0
├── @shelf/jest-mongodb@4.1.7
├── @types/archiver@5.3.2
├── @types/async-retry@1.4.5
├── @types/async@3.2.20
├── @types/correlation-id@2.1.2
├── @types/exceljs@1.3.0
├── @types/express@4.17.17
├── @types/generic-pool@3.8.1
├── @types/jest@29.5.3
├── @types/lodash@4.14.195
├── @types/snowflake-sdk@1.6.13
├── @types/unzipper@0.10.6
├── @typescript-eslint/eslint-plugin@5.62.0
├── @typescript-eslint/parser@5.62.0
├── agenda@5.0.0
├── async-retry@1.3.3
├── async@3.2.4
├── correlation-id@5.0.2
├── dd-trace@4.7.0
├── delay@5.0.0
├── eslint-config-airbnb-typescript@17.1.0
├── eslint-config-prettier@8.8.0
├── eslint-plugin-import@2.27.5
├── eslint-plugin-prettier@4.2.1
├── eslint@8.45.0
├── exceljs@4.3.0
├── express@4.18.2
├── generic-pool@3.9.0
├── is-iso-date@0.0.1
├── jest-mock-extended@3.0.4
├── jest@29.6.1
├── lodash@4.17.21
├── mock-jwks@2.1.1
├── moment@2.29.4
├── mongodb@5.7.0
├── nock@13.3.2
├── node-config-ts@3.3.1
├── nodemon@2.0.22
├── prettier@2.8.8
├── snowflake-sdk@1.6.23
├── ts-custom-error@3.3.1
├── ts-jest@29.1.1
├── typescript@5.1.6
├── unzipper@0.10.14
├── uuid@9.0.0
└── winston@3.10.0

5.Server version:* E.g. 1.90.1
You may get the server version by running a query:
```
7.23.1
```
6. What did you do?
Download data for a report with 250000 rows

7. What did you expect to see?
Download should have completed. 
   What should have happened and what happened instead?
We have a service in production that has been running for long and downloads data for reports(using streaming). For the past few days, we are consistently getting below error.
```
uncaughtException: Cannot read properties of null (reading 'getPeerCertificate')
TypeError: Cannot read properties of null (reading 'getPeerCertificate')
    at TLSSocket.validate (/app/node_modules/snowflake-sdk/lib/agent/socket_util.js:90:36)
    at TLSSocket.emit (node:events:524:35)
    at TLSSocket.emit (node:domain:489:12)
    at TLSSocket.emit (/app/node_modules/dd-trace/packages/datadog-instrumentations/src/net.js:61:25)
    at TLSSocket._finishInit (node:_tls_wrap:1038:8)
    at ssl.onhandshakedone (node:_tls_wrap:824:12)
    at TLSWrap.callbackTrampoline (node:internal/async_hooks:130:17)
    
    Network error. Could not reach S3/Blob.
```

8. Can you set logging to DEBUG and collect the logs?

   https://community.snowflake.com/s/article/How-to-generate-log-file-on-Snowflake-connectors

  e.g
  Add this to get standard output.

```
var snowflake = require('snowflake-sdk');
snowflake.configure(
{
  logLevel: 'trace'
});
```
   
9. What is your Snowflake account identifier, if any? (Optional)

**Comment by sfc-gh-dszmolka** [comment]
hello and thank you for submitting this issue. i agree we could look into the error handling around `getPeerCertificate` but that's an improvement and doesn't solve the underlying error, which seems to happen here: https://github.com/snowflakedb/snowflake-connector-nodejs/blob/v1.6.23/lib/agent/socket_util.js#L90
```javascript
const certChain = socket.ssl.getPeerCertificate(true);
```
when getting the cert chain from the S3 bucket endpoint (where the large query results live), to be validated with OCSP later on.

looking at the [getPeerCertificate docs](https://nodejs.org/api/tls.html#tlssocketgetpeercertificatedetailed), it says
> Returns an object representing the peer's certificate. If the peer does not provide a certificate, an empty object will be returned. **If the socket has been destroyed, null will be returned.**

wonder if that can happen here, since there's a `null` returned.

not sure if we can do anything about it on the driver side, but it would be really helpful if you could please

either:
1. enable TRACING loglevel as required per the bug template (`snowflake.configure({ logLevel: "TRACE" })`)
2. enable DEBUG loglevel for the `urllib` library with envvar `DEBUG=urllib`
3. reproduce the problem and gather relevant logs + share them (if you're not comfortable sharing them here, please [work with Snowflake Support](https://community.snowflake.com/s/article/How-To-Submit-a-Support-Case-in-Snowflake-Lodge) 1:1)

OR:
1. provide a reproduction script, which when run, consistently leads to the issue.
I think this is a bit trickier because for now, I believe this issue is tied to your environment rather than the driver (except the lack of error handling...)

**Comment by hassenc** [comment]
> wonder if that can happen here, since there's a null returned.

The error is saying that it's calling `getPeerCertificate` on a `null` object
`Cannot read properties of null (reading 'getPeerCertificate')`
The issue is that `socket.ssl` is null for some reason, and it's not being checked.

**Comment by kumaresansv** [comment]
[extract-2023-07-20T13_58_01.270Z.csv](https://github.com/snowflakedb/snowflake-connector-nodejs/files/12110372/extract-2023-07-20T13_58_01.270Z.csv)

I have uploaded the logs for you. I have masked few information that should not affect reading the logs.

Thanks
Kumar

**Comment by kumaresansv** [comment]
@sfc-gh-dszmolka When we disable OCSP validation(by setting insecureConnect to true), our service is able to download the large query results(from the Azure blobs). We are currently using it as a temporary workaround. 

What seems strange is that the same codebase fails with OCSP validation enabled.  In the log files above (with OCSP enabled), I do see that the azure blob certificate got validated(Line no 70). So, why would we get the error later when downloading the file.

**Comment by kumaresansv** [comment]
Sure, since we have a workaround, we are ok for now. I will see if I can raise a case with Snowflake support.

**Comment by kumaresansv** [comment]
And thanks for the help.

**Comment by sfc-gh-dszmolka** [comment]
looked into the logs and here's what i think.

> When we disable OCSP validation(by setting insecureConnect to true), our service is able to download the large query results(from the Azure blobs)
Per the logs, the service is also able to download the large query results from Azure blobs, even with OCSP validation enabled.

See `Request#22` - `Request#30` (with the exception to `Request#29` where the issue happens) . E.g. chunks data_0_0_1 (Request#23), data_0_0_2 (Request#24), data_0_0_3 (Request#25) seem to be succesfully downloaded and others too. Of course this is only part of the result set.

> I do see that the azure blob certificate got validated(Line no 70). So, why would we get the error later when downloading the file.
All individual connection get OCSP-validated every time , separately. If for example you have 10 result set chunks and 9 out of 10 passes OCSP validation, the 10th can still fail due to various reasons.

> What seems strange is that the same codebase fails with OCSP validation enabled
It actually makes sense why it works when OCSP validation is disabled. The failing code path is only triggered when OCSP is enabled (default behaviour) and when it's disabled we just return the socket without doing anything else (https://github.com/snowflakedb/snowflake-connector-nodejs/blob/v1.6.23/lib/agent/socket_util.js#L61-L65)

What I think is failing is the connection in Request#29. The DNS resolution looks successful, but when trying to read the certificate chain from the remote end, it fails with the error. As we established earlier, the socket seems to be perceived as `null` in this case. Perhaps the remote end closed it without notifying the client? Or it somehow got destroyed without the driver knowing about it ?

Speaking about the DNS resolution, might or might not be relevant: it seems that in the one single unsuccessful occassion, a different blob storage IP (20.60.180.97) was used than for the rest of the connections (20.209.68.1), but this might be not even true in the context of the whole logfile. I just see a snippet. Probably a red herring for now and we can set aside.

Probably running a packet capture could gather more information about this situation, what exectly happens prior to or during when the socket is null.
❓  Also out of curiosity, does it make any difference if you try to run your application without the DataDog `dd-trace` instrumentation ? Does the exact same issue still happen with OCSP enabled ?

⚫  But if you're up for a little experimenting, we can try taking the same code path as 'OCSP is disabled', so simply returning the socket. If you say it's working with OCSP disabled, I would assume the socket is still null there but the driver is then handling it. So we can try something similar, until the official 'make getting client cert chain faster' improvement is ready.

1. prepated a patch for you which logs an extra line if an exception is encountered during `getPeerCertificate` + returns the socket, similarly how its done in 'OCSP is disabled' path

2. (optional) make a backup of `lib/agent/socket_util.js`

3. copy the patch in `node_modules/snowflake-sdk/lib/agent` and `cd` there

4. apply the patch with `patch socket_util.js < socket_util.js.patch`
Preferably, on a lower level environment if there's one available.

5. disable `insecureConnect` and test if the patch works and you're now seeing the new loglines logged (when the exception happens) and able to download the result sets fully 

Of course this might be a bit different in ephemeral environments :) or maybe you don't want to experiment with this. 
⚫ In this case, you might give a shot trying connection pooling with idle session eviction. Something like this:
```javascript
const pool = snowflake.createPool(
    {
      account: account,
      username: username,
..rest of the connection options..
    },
    {
      evictionRunIntervalMillis: 60000 // default = 0, off
      idleTimeoutMillis: 60000, // default = 30000
      max: 2,
      min: 0,
    },
  );
```
would create a 'pool' with max 2 connections (you'll probably want more, this is just an example), with eviction checks every minute and evicting connections being idle for more than 1 minute.

Closing down idle connections on the driver side can help with situations where the remote end silently closes idle connections after X amount of time. Not sure if it applies to you.

**Comment by kumaresansv** [comment]
That was a lot for me to take in. I really like the patch idea as that will give us a workaround with security in place. I will work on the patch and get it deployed and validated. Any chance the patch could be implemented in future releases.

Regarding connection pooling, we already are evicting idle sessions. Here is the setting that we are using for connecting to Snowflake. We have preFetch set to 1 for better memory management when downloading multiple large query results.

`
      "clientSessionKeepAlive": true,
      "clientSessionKeepAliveHeartbeatFrequency": 900,
      "acquireTimeoutMillis": 5000,
      "evictionRunIntervalMillis": 60000,
      "idleTimeoutMillis": 120000,
      "resultPrefetch": 1 
`

Question
When the library is trying to download large result sets from Azure Blob, does it still check if the session with Snowflake is active and not dropped. Just trying to understand the relation between evicting connections and issue at hand.

**Comment by sfc-gh-dszmolka** [comment]
> Any chance the patch could be implemented in future releases.

There's absolutely no guarantee the quick patch actually works :) It's really just an idea at this point and really hope it works out for you as a stopgap measure. If it does, then we have actual proof it works and can consider including it in later releases. Or figure out something more sophisticated to address the 'read certificate chain' situation and the error handling around it.

> Regarding connection pooling, we already are evicting idle sessions.

Good to know that and thank you for sharing the config snippet. Sadly I don't know from the top of my head if/how we confirm the connection is actually up before using it for large result fetch from the blob. I'll need to look into that, and perhaps set up a reproduction environment too.

**Comment by sfc-gh-dszmolka** [comment]
quick update: spent couple of hours and countless runs on desperately trying to reproduce the issue on node:18 env
```
# node -v
v18.16.1
# grep -i pretty /etc/*release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
```

to no avail. issue could not be reproduced.

then I remembered you're running node:20, spun up a new environment:
```
# node -v
v20.4.0
root@3565d5722fb4:/node# grep -i pretty /etc/*release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
```

and immediately, on the very first run, the issue reproduced. Very interestingly, only with node:20.

Also on the same note I could try the patch myself, it showed it's not working because I made a mistake of letting the `certChain` be `const` and therefore after the try block it becomes undefined. 
Replacing it with `var` allowed the patch to work correctly and all the query results returned, even with OCSP verification.

I'll try to dig further into this delicacy with node:20.

**Comment by sfc-gh-dszmolka** [comment]
I'm still missing tons of details, but found that for when the socket is `null` and the exception happens, when `NODE_DEBUG` is set to `net,tls`, this pattern was observed:
```
{"level":"TRACE","message":"[1:45:21.4521 PM]: https://bnbs3asfcb1stg.blob.core.windows.net/results/01add6bd-0c04-cc87-0000-00036d9a4021_0%2Fmain%2Fdata_0_5_0?..."}
TLS 678: client _init handle? true
NET 678: pipe false null
NET 678: connect: find host bnbs3asfcb1stg.blob.core.windows.net
NET 678: connect: dns options { family: undefined, hints: 32 }
NET 678: connect: autodetecting
TLS 678: client initRead handle? true buffered? false
NET 678: _read - n 16384 isConnecting? true hasHandle? true
NET 678: _read wait for connection
NET 678: connect/multiple: will try the following addresses [
  { address: '20.209.90.161', family: 4 },
  { address: '20.209.111.65', family: 4 },
  { address: '20.209.110.225', family: 4 }
]
>>> NET 678: connect/multiple: attempting to connect to 20.209.90.161:443 (addressType: 4)
NET 678: connect/multiple: setting the attempt timeout to 250 ms
..
{"level":"TRACE","message":"[1:45:21.4521 PM]: https://bnbs3asfcb1stg.blob.core.windows.net/results/01add6bd-0c04-cc87-0000-00036d9a4021_0%2Fmain%2Fdata_0_5_1?..."}
TLS 678: client _init handle? true
NET 678: pipe false null
NET 678: connect: find host bnbs3asfcb1stg.blob.core.windows.net
NET 678: connect: dns options { family: undefined, hints: 32 }
NET 678: connect: autodetecting
TLS 678: client initRead handle? true buffered? false
NET 678: _read - n 16384 isConnecting? true hasHandle? true
NET 678: _read wait for connection
>>> NET 678: connect/multiple: connection to 20.209.90.161:443 timed out
TLS 678: client _init handle? true
NET 678: connect/multiple: attempting to connect to 20.209.111.65:443 (addressType: 4)
NET 678: connect/multiple: setting the attempt timeout to 250 ms
NET 678: connect/multiple: will try the following addresses [
  { address: '20.209.90.161', family: 4 },
  { address: '20.209.111.65', family: 4 },
  { address: '20.209.110.225', family: 4 }
]
NET 678: connect/multiple: attempting to connect to 20.209.90.161:443 (addressType: 4)
NET 678: connect/multiple: setting the attempt timeout to 250 ms
NET 678: connect/multiple: connection attempt to 20.209.90.161:443 completed with status 0
NET 678: afterConnect
TLS 678: client _start handle? true connecting? false requestOCSP? false
NET 678: _read - n 16384 isConnecting? false hasHandle? true
NET 678: Socket._handle.readStart
NET 678: connect/multiple: connection attempt to 20.209.111.65:443 completed with status 0
NET 678: afterConnect
TLS 678: client _start handle? true connecting? false requestOCSP? false
NET 678: _read - n 16384 isConnecting? false hasHandle? true
NET 678: Socket._handle.readStart
NET 678: _read - n 16384 isConnecting? false hasHandle? true
TLS 678: client onhandshakedone
TLS 678: client _finishInit handle? true alpn false servername bnbs3asfcb1stg.blob.core.windows.net
TLS 678: client emit secureConnect. authorized: true
/node/node_modules/snowflake-sdk/lib/agent/socket_util.js:90
      const certChain = socket.ssl.getPeerCertificate(true);
                                   ^

TypeError: Cannot read properties of null (reading 'getPeerCertificate')
    at TLSSocket.validate (/node/node_modules/snowflake-sdk/lib/agent/socket_util.js:90:36)
    at TLSSocket.emit (node:events:524:35)
    at TLSSocket.emit (node:domain:489:12)
    at TLSSocket._finishInit (node:_tls_wrap:1038:8)
    at ssl.onhandshakedone (node:_tls_wrap:824:12)

Node.js v20.4.0
```

so seemingly, the connection is reattempted and successful after the timeout, but in [internalConnectMultipleTimeout](https://github.com/nodejs/node/blob/v20.4.0/lib/net.js#L1678-L1684) where this error message comes from, there is a very suspicious `handle.close()`:
```
function internalConnectMultipleTimeout(context, req, handle) {
  debug('connect/multiple: connection to %s:%s timed out', req.address, req.port);
  req.oncomplete = undefined;
  ArrayPrototypePush(context.errors, createConnectionError(req, UV_ETIMEDOUT));
  handle.close();
  internalConnectMultiple(context);
}
```

to compare, [similar function](https://github.com/nodejs/node/blob/v18.16.1/lib/net.js#L1563-L1566) in node:18:
```
function internalConnectMultipleTimeout(context, req) {
  context[kTimeoutTriggered] = true;
  internalConnectMultiple(context);
}
```
seems to be implemented differently. no `handle.close()`.
regardless, since the issue is very well reproducible on node:20, this pattern is always present in the occassions where the exception occurs, and never present in the connections which go well. consequently, the timeout is always followed by the exception in the repros.

also the `attempt timeout` of `250 ms` seems to come from [here](https://github.com/nodejs/node/blob/v20.4.0/lib/net.js#L132-L133) which again was [implemented differently](https://github.com/nodejs/node/blob/v18.16.1/lib/net.js#L1242) in node:18 inside the `lookupAndConnect` function, really not sure if it makes a difference tho.

but since it seems to be related to `autoSelectFamilyAttemptTimeoutDefault`, as an experiment I tried to run the repro script on node:20 with `--no-network-family-autoselection` and as a result:
* messages with `connect/multi` disappeared from `NET` debug
* also the error itself disappeared. :|

even without the patched Snowflake `socket_util.js`, even with OCSP error, I did not see the error. maybe just a coincidence, but really strange that 50% of the runs reproduce the error, until `--no-network-family-autoselection` is applied, then it doesn't reproduce.

again, this is full of gaps and perhaps even wrong conclusion, but for now at this moment, to me, it looks like this is coming from `node` itself and `snowflake-sdk` just trips up on how it cannot handle the underlying error. 

### to sum up, current options:
1. enhancement in `snowflake-sdk`, to handle such error more intelligently. This ticket will be used to track the enhancement, with no committed timeline.
2. hotfix patch as already provided, for `socket_util.js` which uses the same approach as disabling OCSP, just returns the socket upon exception encountered. (with correcting `const certChain` to `var certChain` inside the `try` block). Reuploading the working version:
[socket_util.js.patch](https://github.com/snowflakedb/snowflake-connector-nodejs/files/12146940/socket_util.js.patch)
3. disabling OCSP with `insecureConnect`
4. using `node:18` where the issue is not reproducible
5. using `node:20` with `--no-network-family-autoselection` (**note**: perhaps this needs further testing in your environment to make sure it doesn't have any unseen impact which didn't come out in the repros, e.g. performance)

**Comment by kumaresansv** [comment]
Thanks. That's a big relief that you are able to reproduce. I had to move to latest version of node to fix some CVE vulnerabilities that were affecting our pipelines. When I get time, I will apply the path and hopefully the issue gets resolved sooner.

**Comment by sfc-gh-dszmolka** [comment]
i found this node:20 bug https://github.com/nodejs/node/issues/48426 and what is described in there looks similar to me what is happening in the repro scenarios. but i also see it's fixed in node:20.4 so i'm baffled why it is still happening for us.

**Comment by APTy** [comment]
Appreciate all the detail in this thread, it's been very insightful 💡 

As an extra data point, we're using `node:18` and see this error occasionally (maybe once per week, hard to reproduce), but we also don't use connection pools yet.

I've applied the patch ([socket_util.js.patch](https://github.com/snowflakedb/snowflake-connector-nodejs/files/12146940/socket_util.js.patch) --> _note for future readers there are two patch links in this thread, and this is the correct one_) and will see how it goes.

My take is: regardless of the underlying upstream issue, it seems that there are various places where the socket_util.js logic can fail, (i.e. `socket` becomes `null`, and I'd expect even `getPeerCertificate()` could throw) and `snowflake-sdk` should handle those errors and not cause a hard process crash.

🙏 🙏 🙏

**Comment by sfc-gh-dszmolka** [comment]
i feel like the `socket hang up` might be a different error. people see it even without seeing the error which is handled in this Issue, even with different versions of `node` other than v20. 

this issue is addressing the `TypeError: Cannot read properties of null (reading 'getPeerCertificate'` error, and thank you for reminding me of the wrong patch version , it has been removed from the relevant post, and only the working left.

**Comment by manishpragma** [comment]
I have faced this issue on node@20.5! Downgraded to node@18 and working fine for me!!

**Comment by sfc-gh-dszmolka** [comment]
we found a possible (probable) cause for this issue, which is addressed in PR #645 . Depending on how the review procedure goes and when the PR gets merged; it'll either be part of this (September) release, or the next (October). Will keep this Issue updated.

**Comment by sfc-gh-dszmolka** [comment]
PR #645 merged and will be released with the next driver release, towards end of September.

**Comment by sfc-gh-dszmolka** [comment]
fix is now out with release [1.9.0](https://github.com/snowflakedb/snowflake-connector-nodejs/releases/tag/v1.9.0) , [available on npm](https://www.npmjs.com/package/snowflake-sdk/v/1.9.0) as well

**Comment by ZeRego** [comment]
@sfc-gh-dszmolka Just wanna let you know that we have experienced this error on node 20.8.0 when using snowflake-sdk 1.9.0. 

We haven't been able to reproduce this locally on a controlled environment.

**Comment by sfc-gh-dszmolka** [comment]
Thank you - please once you manage to reproduce it, open a new issue or post it here. This one seems to be resolved with 1.9.0 per the tests, but if there's any other issue left or any other scenario where it's still reproducible, we'd like to take a look. Thank you in advance !
